File: 数值算法Ⅰ/Homework & Exam/数值算法 01/HW01_P7_validation.c
#include <stdio.h>
#include <float.h>

void test_float() {
    float H = 1.0f;
    int n = 2;
    while (1) {
        float term = 1.0f / (float)n;
        float newH = H + term;
        if (newH == H) break;
        H = newH;
        n++;
    }
    printf("Single precision stops at n = %d\n", n);
    printf("H_n = %.8f\n", H);
}

int main() {
    test_float();
    return 0;
}

File: 数值算法Ⅰ/Homework & Exam/数值算法 01/HW01_P7_validation.m
% 验证调和级数在单精度浮点数运算中的"收敛"现象
H_single = single(0.0);
n_single = single(1.0);   % 用 single 而不是 int32

while true
    if H_single == H_single + 1.0/n_single
        break;
    end
    H_single = H_single + 1.0./n_single;
    n_single = n_single + 1.0;
end

fprintf('Single precision stops at n = %.0f\n', n_single);
fprintf('H_n = %.8f\n', H_single);